src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -654,76 +635,52 @@ public new CSharpCompilation AddSyntaxTrees(IEnumerable<SyntaxTree> trees)
+            int i = 0;
+            foreach (var tree in trees.Cast<CSharpSyntaxTree>())
             {
-                builder.AddRange(this.SyntaxTrees);
-
-                bool referenceDirectivesChanged = false;
-                var oldTreeCount = this.SyntaxTrees.Length;
-                var ordinalMap = this.syntaxTreeOrdinalMap;
-                var declMap = _rootNamespaces;
-                var declTable = _declarationTable;
-                int i = 0;
-                foreach (var tree in trees.Cast<CSharpSyntaxTree>())
+                if (tree == null)
                 {
-                    if (tree == null)
-                    {
-                        throw new ArgumentNullException("trees[" + i + "]");
-                    }
-
-                    if (!tree.HasCompilationUnitRoot)
-                    {
-                        throw new ArgumentException(String.Format(CSharpResources.TreeMustHaveARootNodeWith, i));
-                    }
-
-                    if (declMap.ContainsKey(tree))
-                    {
-                        throw new ArgumentException(CSharpResources.SyntaxTreeAlreadyPresent, String.Format(CSharpResources.Trees0, i));
-                    }
-
-                    if (IsSubmission && tree.Options.Kind == SourceCodeKind.Regular)
-                    {
-                        throw new ArgumentException(CSharpResources.SubmissionCanOnlyInclude, String.Format(CSharpResources.Trees0, i));
-                    }
+                    throw new ArgumentNullException($"{nameof(trees)}[{i}]");
+                }
 
-                    AddSyntaxTreeToDeclarationMapAndTable(tree, _options, IsSubmission, ref declMap, ref declTable, ref referenceDirectivesChanged);
-                    builder.Add(tree);
-                    ordinalMap = ordinalMap.Add(tree, oldTreeCount + i);
+                if (!tree.HasCompilationUnitRoot)
+                {
+                    throw new ArgumentException(String.Format(CSharpResources.TreeMustHaveARootNodeWith, i));
+                }
 
-                    i++;
+                if (externalSyntaxTrees.Contains(tree))
+                {
+                    throw new ArgumentException(CSharpResources.SyntaxTreeAlreadyPresent, String.Format(CSharpResources.Trees0, i));
                 }
 
-                if (IsSubmission && declMap.Count > 1)
+                if (this.IsSubmission && tree.Options.Kind == SourceCodeKind.Regular)
                 {
-                    throw new ArgumentException(CSharpResources.SubmissionCanHaveAtMostOne, "trees");
+                    throw new ArgumentException(CSharpResources.SubmissionCanOnlyInclude, String.Format(CSharpResources.Trees0, i));
                 }
 
-                return UpdateSyntaxTrees(builder.ToImmutable(), ordinalMap, declMap, declTable, referenceDirectivesChanged);
+                externalSyntaxTrees.Add(tree);
+                reuseReferenceManager &= !tree.HasReferenceDirectives && !tree.HasLoadDirectives();
```

Comment:
Is it worth adding a HasReferenceOrLoadDirectives method? There are multiple places where we check for both.